src/os_common/os_fhandle.c

```diff
@@ -355,6 +357,108 @@ __wt_close(WT_SESSION_IMPL *session, WT_FH **fhp)
 	return (__handle_close(session, fh, true));
 }
 
+/*
+ * __wt_fsync_background_chk --
+ *	Return if background fsync is supported.
+ */
+bool
+__wt_fsync_background_chk(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_FH *fh, *fhnext;
+	WT_FILE_HANDLE *handle;
+	bool supported;
+
+	conn = S2C(session);
+	supported = true;
+	__wt_spin_lock(session, &conn->fh_lock);
+	/*
+	 * Look for the first data file handle and see if
+	 * the fsync nowait function is supported.
+	 */
+	TAILQ_FOREACH_SAFE(fh, &conn->fhqh, q, fhnext) {
+		handle = fh->handle;
+		if (fh->file_type != WT_FS_OPEN_FILE_TYPE_DATA)
+			continue;
+		/*
+		 * If we don't have a function, return false, otherwise
+		 * return true. In any case, we are done with the loop.
+		 */
+		if (handle->fh_sync_nowait == NULL)
+			supported = false;
+		break;
+	}
+	__wt_spin_unlock(session, &conn->fh_lock);
+	return (supported);
+}
+
+/*
+ * __wt_fsync_background --
+ *	POSIX background fsync for all dirty file handles.
+ */
+int
+__wt_fsync_background(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_FH *fh;
+	WT_FILE_HANDLE *handle;
+	uint64_t now;
+
+	conn = S2C(session);
+	__wt_spin_lock(session, &conn->fh_lock);
+	TAILQ_FOREACH(fh, &conn->fhqh, q) {
+		handle = fh->handle;
+		WT_STAT_CONN_INCR(session, fsync_all_fh_total);
+		if (handle->fh_sync_nowait == NULL ||
+		    fh->written < WT_CAPACITY_FILE_THRESHOLD)
+			continue;
+		/* Only sync data files. */
+		if (fh->file_type != WT_FS_OPEN_FILE_TYPE_DATA)
+			continue;
+		now = __wt_clock(session);
+		if (fh->last_sync == 0 ||
+		    WT_CLOCKDIFF_SEC(now, fh->last_sync) > 0) {
+			/*
+			 * Increment our ref count on the current handle.
+			 * That way it is guaranteed valid when we
+			 * lock again after the fsync.
+			 */
+			++fh->ref;
+			__wt_spin_unlock(session, &conn->fh_lock);
+			ret = __wt_fsync(session, fh, false);
+			/*
+			 * Although we send in the false flag to indicate a
+			 * non-blocking background fsync, there is no guarantee
+			 * that it doesn't block. If we wanted to detect if it
+			 * is blocking, adding a clock call and checking the
+			 * time would be done here.
+			 */
+			/*
+			 * Even if we get an error, we still need to lock and
+			 * decrement our reference counts.
+			 */
+			if (ret == 0) {
+				WT_STAT_CONN_INCR(session, fsync_all_fh);
+				fh->last_sync = now;
+				fh->written = 0;
+			}
+			__wt_spin_lock(session, &conn->fh_lock);
+			--fh->ref;
+			WT_ERR(ret);
+			/*
+			 * The last open reference to the file handle may
+			 * have been released while we dropped the lock.
+			 * If so, we're responsible for closing.
+			 */
+			if (fh->ref == 0)
+				return (__handle_close(session, fh, true));
```

Comment:
If `__wt_fsync` fails, we don't panic, but we potentially leak a file handle. Any, why are we returning out of the loop here if we're closing a  handle, aren't there other handles to sync (plus, we're returning while holding the spin lock).

I'd suggest:
```
*** os_fhandle.c        Wed Feb  6 15:08:45 2019
--- -   Wed Feb  6 15:23:46 2019
***************
*** 434,458 ****
                         * is blocking, adding a clock call and checking the
                         * time would be done here.
                         */
-                       /*
-                        * Even if we get an error, we still need to lock and
-                        * decrement our reference counts.
-                        */
                        if (ret == 0) {
                                WT_STAT_CONN_INCR(session, fsync_all_fh);
                                fh->last_sync = now;
                                fh->written = 0;
                        }
!                       __wt_spin_lock(session, &conn->fh_lock);
!                       --fh->ref;
!                       WT_ERR(ret);
                        /*
!                        * The last open reference to the file handle may
!                        * have been released while we dropped the lock.
!                        * If so, we're responsible for closing.
                         */
                        if (fh->ref == 0)
!                               return (__handle_close(session, fh, true));
                }
        }
  err:  __wt_spin_unlock(session, &conn->fh_lock);
--- 434,460 ----
                         * is blocking, adding a clock call and checking the
                         * time would be done here.
                         */
                        if (ret == 0) {
                                WT_STAT_CONN_INCR(session, fsync_all_fh);
                                fh->last_sync = now;
                                fh->written = 0;
                        }
!
                        /*
!                        *
!                        * Even if we get an error, we still need to lock and
!                        * decrement our reference counts.
!                        *
!                        * The last open reference to the file handle may have
!                        * been released while we dropped the lock. If so, we're

!                        * responsible for closing.
                         */
+                       __wt_spin_lock(session, &conn->fh_lock);
+                       --fh->ref;
                        if (fh->ref == 0)
!                               WT_TRET(__handle_close(session, fh, true));
!
!                       WT_ERR(ret);
                }
        }
  err:  __wt_spin_unlock(session, &conn->fh_lock);
```

Reply:
@keithbostic, I've just pushed some changes that are a little different than you suggest.  `__handle_close` when called with `true`, releases the `fh_lock`.  I assume we want it to do that, as the handle close may take some time.  So we need to reacquire the lock.  Also, I followed through on your thought that we shouldn't exit the loop on account of any single failure.

Reply:
@ddanderson, I think this is close, but I think you'll now need `TAILQ_FOREACH_SAFE`, because we're using `fh` after it's been freed, aren't we?

Reply:
@keithbostic, yes you're right.  If we have a temp variable, we'll need to increase its ref count also so it doesn't get deleted out from under us.  I think we did that in a past version of this code.

Reply:
@keithbostic, I've pushed a fix for your consideration.

Reply:
I think we can still get into trouble:

First, if `fhtmp` goes to 0, but that handle isn't a file handle that we will flush on the next iteration of the outer loop, we `continue` the outer loop instead of falling into the check-for-handle-close code, so it won't get closed.

Second, once we decrement the `fhtmp` reference count, can't the close of the `fh` handle (discarding the spinlock), allow `fhtmp` to be closed out from under us?

Also, I don't think it matters, but checking for `fh->ref > 0` looks wrong to me. We're holding the spinlock, is it possible for a handle to be on that list with a zero ref count?

I can't see any way to make the existing loop work without a bunch of goto's and special cases.

I've pushed a branch in 92e175a for your consideration.

I don't think it's any slower, and it seems cleaner and less fiddly to me -- we've had enough problems getting this correct I'd like to avoid anything complicated. Feel free to simply revert the change if you don't like it, or I'm misunderstanding the code, or you just have something you like better!

Reply:
Thanks to both of you for working through this. LGTM.